### PR TITLE
refact: extract adaptor from protocols

### PR DESCRIPTION
**Pull request by asterwyx**
+ fix cannot generate local protocols
 + extract adaptor from protocols

### CODE REVIEW

File: .reuse/dep5 [hunk 1]
```diff
@@ -13,6 +13,11 @@ Files: .gitignore .gitmodules
 Copyright: None
 License: CC0-1.0
 
+# protocol and data files
+Files: *.xml
+Copyright: None
+License: CC0-1.0
```
Comment (wineee) [inline .reuse/dep5:19]: No, it's MIT

**Review by wineee — CHANGES_REQUESTED** [review]

**Comment by wineee** [comment]
https://github.com/qt/qtbase/blob/785a5e6eb4745dc3219cc33dc5ea2e671a40b5f3/src/network/access/qhttpheaders.h#L164

File: src/server/qtquick/private/wquicktextinputv1.cpp [hunk 2]
```diff
@@ -199,377 +53,296 @@ QList<WQuickTextInputV1 *> WQuickTextInputManagerV1::textInputs() const
...
 
-void text_input_handle_commit_state(struct wl_client *client,
-                                    struct wl_resource *resource,
-                                    uint32_t serial)
+im::ContentPurpose WTextInputV1Adaptor::contentPurpose() const
 {
-    ws_text_input_v1 *text_input = text_input_from_resource(resource);
-    text_input->current_serial = serial;
-    wl_signal_emit_mutable(&text_input->events.commit_state, text_input);
+    const QMap<WQuickTextInputV1::ContentPurpose, IM::ContentPurpose> purposeMap = {
```
[the remark this answers is not in the document]
Comment (zccrs) [reply src/server/qtquick/private/wquicktextinputv1.cpp:281]: Maybe QHash is better for this cause.

File: src/server/qtquick/winputmethodhelper.h [hunk 3]
```diff
@@ -0,0 +1,200 @@
+// Copyright (C) 2023 Yixue Wang <wangyixue@deepin.org>.
```
Comment (zccrs) [inline src/server/qtquick/winputmethodhelper.h:1]: Why isn't private header?

File: src/server/kernel/wtextinputv1.h [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+WAYLIB_SERVER_BEGIN_NAMESPACE
+struct ws_text_input_v1;
+struct ws_text_input_manager_v1;
+class WSurface;
+class WSeat;
+class WTextInputV1;
+class WTextInputManagerV1Private;
+class WTextInputV1Private;
+class WTextInputManagerV1 : public QObject, public WObject
```
Comment (zccrs) [inline src/server/kernel/wtextinputv1.h:18]: WAYLIB_SERVER_EXPORT

**Review by zccrs — CHANGES_REQUESTED** [review]

File: src/server/kernel/wtextinputv1.h [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    WSeat *seat() const;
+    WSurface *focusedSurface() const;
+    QString surroundingText() const;
+    uint surroundingCursor() const;
+    uint surroundingAnchor() const;
+    uint contentHint() const;
+    uint contentPurpose() const;
+    QRect cursorRectangle() const;
+    QString preferredLanguage() const;
+    wl_client *waylandClient() const;
```
Comment (zccrs) [inline src/server/kernel/wtextinputv1.h:56]: WObjectPrivate has been waylandClient function, so this public interface can add to WObject?

File: src/server/kernel/wtextinputv1.h [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+Q_SIGNALS:
+    void activate(WSeat *seat, WSurface *surface);
+    void deactivate(WSeat *seat);
+    void showInputPanel();
+    void hideInputPanel();
+    void reset();
+    void commitState();
+    void invokeAction(uint button, uint index);
+    void beforeDestroy(WTextInputV1 *self);
```
Comment (zccrs) [inline src/server/kernel/wtextinputv1.h:66]: `self` is unnecessary.

**Review by zccrs — CHANGES_REQUESTED** [review]

File: src/server/kernel/wtextinputv1.h (hunk 5 above)
Comment (asterwyx) [reply src/server/kernel/wtextinputv1.h:56]: I don't know if this interface is necessary in other protocol entities. But if export this in base class. That might be convenient.

Comment (asterwyx) [reply src/server/kernel/wtextinputv1.h]: For my design, exporting waylandClient might add complexity, I will move all client's checks to private classes.

Comment (asterwyx) [reply src/server/kernel/wtextinputv1.h]: Done.

File: src/server/qtquick/private/wquickinputmethodv2_p.h [hunk 7]
```diff
@@ -209,18 +145,10 @@ public Q_SLOTS:
...
 Q_SIGNALS:
-    void commit();
-    void newPopupSurface(WQuickInputPopupSurfaceV2 *surface);
-    void newKeyboardGrab(WQuickInputMethodKeyboardGrabV2 *keyboardGrab);
-
-private:
-    explicit WQuickInputMethodV2(QW_NAMESPACE::QWInputMethodV2 *handle, QObject *parent = nullptr);
-    friend class WQuickInputMethodManagerV2;
+    void committed();
+    void newPopupSurface(QWLRoots::QWInputPopupSurfaceV2 *surface);
```
Comment (zccrs) [inline src/server/qtquick/private/wquickinputmethodv2_p.h:151]: Use `QW_NAMESPACE`

File: src/server/qtquick/private/wquicktextinputv1.cpp [hunk 8]
```diff
@@ -164,217 +49,378 @@ class WQuickTextInputManagerV1Private : public WObjectPrivate
...
     wlr_surface *focused_surface = wlr_surface_from_resource(surface);
+    bool focus_changed = (text_input->focused_surface == focused_surface);
     wl_signal_add(&focused_surface->events.destroy, &text_input->surface_destroy);
-    // FIXME: Should we use a pending state?
     text_input->focused_surface = focused_surface;
-    wl_signal_emit_mutable(&text_input->events.activate, text_input);
-    // TODO: As soon as activated, compositor should send enter to client.
+    // FIXME: Does the order matter?
+    if (seat_changed) {
+        wl_signal_emit(&text_input->events.seat_changed, text_input);
```
Comment (zccrs) [inline src/server/qtquick/private/wquicktextinputv1.cpp:383]: wl_signal_emit_mutable ?

File: src/server/qtquick/private/wquicktextinputv1.cpp [hunk 9]
```diff
@@ -164,217 +49,378 @@ class WQuickTextInputManagerV1Private : public WObjectPrivate
...
+    }
 }
 
 void text_input_handle_show_input_panel(struct wl_client *client,
                                         struct wl_resource *resource)
 {
     ws_text_input_v1 *text_input = text_input_from_resource(resource);
-    wl_signal_emit_mutable(&text_input->events.show_input_panel, text_input);
+    wl_signal_emit(&text_input->events.show_input_panel, text_input);
```
Comment (zccrs) [inline src/server/qtquick/private/wquicktextinputv1.cpp:408]: Why remove `_mutable`?

File: src/server/qtquick/private/wquickvirtualkeyboardv1_p.h [hunk 10]
```diff
@@ -24,36 +26,28 @@ class WInputDevice;
...
 
 Q_SIGNALS:
-    void newVirtualKeyboard(WQuickVirtualKeyboardV1 *vk);
+    void newVirtualKeyboard(QW_NAMESPACE::QWVirtualKeyboardV1 *virtualKeyboard);
 
 private:
     void create() override;
 };
 
+
```
Comment (zccrs) [inline src/server/qtquick/private/wquickvirtualkeyboardv1_p.h:52]: remove

**Review by zccrs — CHANGES_REQUESTED** [review]

**Review by zccrs — APPROVED** [review]

**Pull request merged**